daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/DamlBigDecimal.scala

```diff
@@ -0,0 +1,141 @@
+    if (newFlags contains Invalid)
+      DamlBigDecimal(JBigDec.ONE, newFlags)
+    else if (denominator.bigDecimal == JBigDec.ZERO)
+      DamlBigDecimal(JBigDec.ONE, newFlags + Invalid)
+    else if (numerator.bigDecimal == JBigDec.ZERO)
+      DamlBigDecimal(JBigDec.ZERO, newFlags)
+    else if (maxPrecision > Integer.MAX_VALUE || maxPrecision < 1)
+      DamlBigDecimal(JBigDec.ONE, newFlags + Invalid)
+    else
+      try {
+        val mc = new MathContext(maxPrecision.toInt, RoundingMode.valueOf(roundingMode))
+        val d = numerator.bigDecimal.divide(denominator.bigDecimal, mc)
+        if (d == JBigDec.ZERO)
+          DamlBigDecimal(d, newFlags + Underflow)
+        else
+          DamlBigDecimal(d, newFlags)
+      } catch {
+        case _: ArithmeticException =>
+          DamlBigDecimal(JBigDec.ONE, Set.empty + Invalid)
+      }
+  }
+
+  def divMod(maxPrecision: Long, roundingMode: String, numerator: DamlBigDecimal, denominator: DamlBigDecimal): (DamlBigDecimal, DamlBigDecimal) = {
+    val newFlags = numerator.dbdFlags union denominator.dbdFlags
+    if (newFlags contains Invalid)
+      (DamlBigDecimal(JBigDec.ONE, newFlags), DamlBigDecimal(JBigDec.ONE, newFlags))
+    else if (denominator.bigDecimal == JBigDec.ZERO)
+      (DamlBigDecimal(JBigDec.ONE, newFlags + Invalid), DamlBigDecimal(JBigDec.ONE, newFlags + Invalid))
+    else if (numerator.bigDecimal == JBigDec.ZERO)
+      (DamlBigDecimal(JBigDec.ZERO, newFlags), DamlBigDecimal(JBigDec.ZERO, newFlags))
+    else if (maxPrecision > Integer.MAX_VALUE || maxPrecision < 1)
+      (DamlBigDecimal(JBigDec.ONE, newFlags + Invalid), DamlBigDecimal(JBigDec.ONE, newFlags + Invalid))
+    else
+      try {
+        val mc = new MathContext(maxPrecision.toInt, RoundingMode.valueOf(roundingMode))
+        val d = numerator.bigDecimal.divideAndRemainder(denominator.bigDecimal, mc)
+        (DamlBigDecimal(d(0), newFlags), DamlBigDecimal(d(1), newFlags))
+      } catch {
+        case _: ArithmeticException =>
+          (DamlBigDecimal(JBigDec.ONE, newFlags + Invalid), DamlBigDecimal(JBigDec.ONE, newFlags + Invalid))
+      }
+  }
+
+  def compare(lhs: DamlBigDecimal, rhs: DamlBigDecimal): Int =
+    (lhs.dbdFlags contains Invalid, rhs.dbdFlags contains Invalid) match {
+      case (true, true) => 0
+      case (true, false) => -1
+      case (false, true) => 1
+      case (false, false) => lhs.bigDecimal compareTo rhs.bigDecimal
+    }
+
+  val zero: DamlBigDecimal = DamlBigDecimal(JBigDec.ZERO, Set.empty)
+
+  val one: DamlBigDecimal = DamlBigDecimal(JBigDec.ONE, Set.empty)
+}
+
+
+sealed trait DBDFlag
+
+// Number was subject to rounding before
+case object Inexact extends DBDFlag
+
+// Number resulted from an illegal operations (e.g., 1 / 0)
+// or was computed using an already-invalid number
+case object Invalid extends DBDFlag
+
+// Number is too small to represent (but not actually zero)
+case object Underflow extends DBDFlag
```

Comment:
We had discussions about simplifying this to a single Boolean (validity flag)

Reply:
I think there are several questions here:

1. Do we want to provide a way to handle errors? Seems sensible to me. What I don’t understand why this is crucial for `BigDecimal` but not for something like `Int64` overflows? Is it just because it’s easier to write your own wrappers for the latter that handle this?
2. How do we expose these failures in DAML? I don’t have a good answer for that.
3. What is the default behavior? This is somewhat similar to `1`. If `Int64` overflows default to aborting the transaction I don’t see any reason why `BigDecimal` overflows should not default to aborting the transaction.
4. How do we differentiate between the default and non-default behavior? This somewhat depends on 2. If we expose it as something like `Either Error BigDecimal` as the result value of each operation, then you need separate operations for those anyway and you cannot use `(+)`. If you add it as a flag in `BigDecimal` itself, then you probably want to be able to use `(+)` in all cases. You could either make it a type parameter here or put it on the value-level. In the latter case, you have to define what happens if the flag is different in the arguments to `(+)` (I suppose aborting the transaction isn’t going to satisfy you here either).

Reply:
So, about 1) we should first note that our design principle was "no land mines". Maybe this is not your opinion, but we think that `BigDecimal` should never abort a transaction.
It would be great if `Int64` never overflowed either - it is just less likely (than with `Numeric n`) given the available range/size to work with but it is equally bad for writing a safe application.
For the way failures are exposed (2, and maybe 4) , note that any `BigDecimal` must turn into a Numeric before you can do anything persistent with it. This will yield a `None` instead of a value, so error handling will always be necessary, i.e. the `Optional` return type exposes the failure. For the use case in point, it is not relevant whether the result value is simply too large to fit into the target `Numeric n` or (much less likely) an overflow has happened in some intermediate result.
I hope this clarifies our thinking a bit.

Reply:
> So, about 1) we should first note that our design principle was "no land mines". Maybe this is not your opinion, but we think that `BigDecimal` should never abort a transaction.

I think our misunderstanding comes from the fact that I consider an aborted transaction a reasonable response to invalid inputs. That’s how the standard library is written, it is how all of our examples are written (they rely on `assert` very heavily), …. As I said, I’m aware that your requirements are different here and I’m happy to try and accomodate them. However, I think it is confusing to single out `BigDecimal` here. Either we stick to the behavior we have atm in the standard library and our examples and the default behavior of `BigDecimal` should match that (while still giving you the option of using something that lets you handle the error) or we try to change the general design of the standard library and the examples which tbh I don’t see happening any time soon.